File: src/apps/bench/rustbenchs/src/btreap.rs
/*
 * Copyright (C) 2018 Nils Asmussen <nils@os.inf.tu-dresden.de>
 * Economic rights: Technische Universitaet Dresden (Germany)
 *
 * Copyright (C) 2019-2021 Nils Asmussen, Barkhausen Institut
 *
 * This file is part of M3 (Microkernel-based SysteM for Heterogeneous Manycores).
 *
 * M3 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * M3 is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details.
 */

use m3::col::Treap;
use m3::test;
use m3::time::{CycleInstant, Profiler, Runner};
use m3::{wv_assert_eq, wv_perf, wv_run_test};

pub fn run(t: &mut dyn test::WvTester) {
    wv_run_test!(t, insert);
    wv_run_test!(t, find);
    wv_run_test!(t, clear);
}

fn insert() {
    let mut prof = Profiler::default().repeats(100).warmup(100);

    #[derive(Default)]
    struct BTreeTester(Treap<u32, u32>);

    impl Runner for BTreeTester {
        fn pre(&mut self) {
            self.0.clear();
        }

        fn run(&mut self) {
            for i in 0..10 {
                self.0.insert(i, i);
            }
        }
    }

    wv_perf!(
        "Inserting 10 elements",
        prof.runner::<CycleInstant, _>(&mut BTreeTester::default())
    );
}

fn find() {
    let mut prof = Profiler::default().repeats(100).warmup(50);

    #[derive(Default)]
    struct BTreeTester(Treap<u32, u32>);

    impl Runner for BTreeTester {
        fn pre(&mut self) {
            for i in 0..10 {
                self.0.insert(i, i);
            }
        }

        fn run(&mut self) {
            for i in 0..10 {
                let val = self.0.get(&i);
                wv_assert_eq!(val, Some(&i));
            }
        }

        fn post(&mut self) {
            self.0.clear();
        }
    }

    wv_perf!(
        "Searching for 10 elements",
        prof.runner::<CycleInstant, _>(&mut BTreeTester::default())
    );
}

fn clear() {
    let mut prof = Profiler::default().repeats(100).warmup(100);

    #[derive(Default)]
    struct BTreeTester(Treap<u32, u32>);

    impl Runner for BTreeTester {
        fn pre(&mut self) {
            for i in 0..10 {
                self.0.insert(i, i);
            }
        }

        fn run(&mut self) {
            self.0.clear();
        }
    }

    wv_perf!(
        "Removing 10-element list",
        prof.runner::<CycleInstant, _>(&mut BTreeTester::default())
    );
}

File: src/apps/bench/cppbenchs/benchs/btreap.cc
/*
 * Copyright (C) 2018 Nils Asmussen <nils@os.inf.tu-dresden.de>
 * Economic rights: Technische Universitaet Dresden (Germany)
 *
 * Copyright (C) 2019-2021 Nils Asmussen, Barkhausen Institut
 *
 * This file is part of M3 (Microkernel-based SysteM for Heterogeneous Manycores).
 *
 * M3 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * M3 is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details.
 */

#include <base/Common.h>
#include <base/Panic.h>
#include <base/col/Treap.h>
#include <base/time/Profile.h>

#include <m3/Test.h>

#include "../cppbenchs.h"

using namespace m3;

struct MyTItem : public TreapNode<MyTItem, uint32_t> {
    explicit MyTItem(uint32_t _val) : TreapNode(_val), val(_val) {
    }

    uint32_t val;
};

NOINLINE static void insert() {
    struct TreapInsertRunner : public Runner {
        void run() override {
            for(uint32_t i = 0; i < 10; ++i) {
                treap.insert(new MyTItem(i));
            }
        }
        void post() override {
            MyTItem *item;
            while((item = treap.remove_root()) != nullptr) {
                delete item;
            }
        }

        Treap<MyTItem> treap;
    };

    Profile pr(100, 100);
    TreapInsertRunner runner;
    WVPERF("inserting 10-elements", pr.runner<CycleInstant>(runner));
}

NOINLINE static void find() {
    struct TreapSearchRunner : public Runner {
        void pre() override {
            for(uint32_t i = 0; i < 10; ++i) {
                treap.insert(new MyTItem(i));
            }
        }
        void run() override {
            for(uint32_t i = 0; i < 10; ++i) {
                MyTItem *item = treap.find(i);
                if(!item || item->val != i)
                    panic("Test failed: {} != {}"_cf, item ? item->val : 0, i);
            }
        }
        void post() override {
            MyTItem *item;
            while((item = treap.remove_root()) != nullptr) {
                delete item;
            }
        }

        Treap<MyTItem> treap;
    };

    Profile pr(100, 50);
    TreapSearchRunner runner;
    WVPERF("searching 10-elements", pr.runner<CycleInstant>(runner));
}

NOINLINE static void clear() {
    struct TreapClearRunner : public Runner {
        void pre() override {
            for(uint32_t i = 0; i < 10; ++i) {
                treap.insert(new MyTItem(i));
            }
        }
        void run() override {
            MyTItem *item;
            while((item = treap.remove_root()) != nullptr) {
                delete item;
            }
        }

        Treap<MyTItem> treap;
    };

    Profile pr(100, 100);
    TreapClearRunner runner;
    WVPERF("removing 10-elements", pr.runner<CycleInstant>(runner));
}

void btreap() {
    RUN_BENCH(insert);
    RUN_BENCH(find);
    RUN_BENCH(clear);
}
